### ISSUE TITLE
v2.2.0 does not read or write tags anymore

### ISSUE THREAD

**Issue by tvdeyen** [pat/gutentag#50]
At least for us in [AlchemyCMS](https://github.com/AlchemyCMS/alchemy_cms) the 2.2.0 release and especially 38b31f3bbce8834ee2ab6e9ac14a24765a1cde5f broke the tagging feature.

See the [passing build on Travis](https://travis-ci.org/tvdeyen/alchemy_cms/builds/349191948) for v2.1.0
See the [broken build on Travis](https://travis-ci.org/tvdeyen/alchemy_cms/builds/349214334) for v2.2.0

### How to replicate locally

1. Clone [the Alchemy branch](https://github.com/tvdeyen/alchemy_cms/tree/gutentag-bug) that shows the bug

```
git clone https://github.com/tvdeyen/alchemy_cms
git checkout -b gutentag-bug origin/gutentag-bug
bundle install
cd spec/dummy
bin/rake db:setup
bin/rails server
```

2. Navigate to https://localhost:3000/admin/locations
3. Create an event location and set a tag
![alchemy cms - locations 2018-03-05 10-47-11](https://user-images.githubusercontent.com/42868/36969115-dc223db4-2064-11e8-976d-6eeb1aee8cb2.png)
4. See the tag appearing in the tag list on the right and under https://localhost:3000/admin/tags
![alchemy cms - locations 2018-03-05 10-56-58](https://user-images.githubusercontent.com/42868/36969135-e8c98a54-2064-11e8-8727-e097f3016835.png)
5. Edit the event location and see that the tag does not appear in the tag list
![alchemy cms - locations 2018-03-05 10-57-07](https://user-images.githubusercontent.com/42868/36969146-f22eb4e8-2064-11e8-830e-09d51327ce66.png)
6. Set another tag
![alchemy cms - locations 2018-03-05 11-01-30](https://user-images.githubusercontent.com/42868/36969156-f84ae586-2064-11e8-90b5-2ea516ab71ce.png)
7. See only the last tag appearing in the tag list on the right
![alchemy cms - locations 2018-03-05 11-01-43](https://user-images.githubusercontent.com/42868/36969164-017b6f90-2065-11e8-9978-3d37cabdc4f5.png)
8. The previous tag is still existing as you can see at https://localhost:3000/admin/tags
![alchemy cms - tags 2018-03-05 11-01-53](https://user-images.githubusercontent.com/42868/36969193-23486f92-2065-11e8-9dfa-75a91569420e.png)

### Investigation

We have a taggable module that is included in all models that need taggings.

```rb
# lib/alchemy/taggable.rb
module Alchemy
  # Gutentag interface compatibility module
  # Include this module to add tagging support to your model.
  module Taggable
    SPLITTER_REGEX = /,\s*/

    def self.included(base)
      Gutentag::ActiveRecord.call base
      base.extend ClassMethods
    end

    def tag_names=(names)
      if names.is_a? String
        names = names.split(SPLITTER_REGEX)
      end
      super(names)
    end

    module ClassMethods
      # Find all records matching all of the given tags.
      # Separate multiple tags by comma.
      def tagged_with(names)
        if names.is_a? String
          names = names.split(SPLITTER_REGEX)
        end
        super(names: names, match: :all)
      end

      # Returns all unique tags
      def tag_counts
        Gutentag::Tag.distinct.joins(:taggings)
          .where(gutentag_taggings: {taggable_type: name})
      end
    end
  end
end
```

It is very basic and only adds some convenient methods to the classes.

First and foremost we need to be able to set tags from a string separated by comma, so we made this change to the setter:

```rb
# lib/alchemy/taggable.rb:12
def tag_names=(names)
  if names.is_a? String
    names = names.split(SPLITTER_REGEX)
  end
  super(names)
end
```

This worked until we updated to 2.2.0. If we remove the calls to `super` in the `tag_names` attribute reader:

```rb
# lib/gutentag/active_record/modern_instance_methods.rb:9
def tag_names
  self.tag_names = tags.pluck(:name)
end
```

it works again. I actually don't see a need to call `super` here. What was the reason for this?

Unfortunately I can't get the writing of tags working with removing calls to `super`. The tags stay the same as before. You can't change them anymore.

It even is broken if we introduce our own writer:

```rb
# lib/alchemy/taggable.rb:12
def tag_list=(names)
  if names.is_a? String
    names = names.split(SPLITTER_REGEX)
  end
  self.tag_names = names
end
```

I am not sure what I am doing wrong here. 

Thanks for the great library and you help is very much appreciated.

**Comment by pat** [comment]
I'm struggling a little to get it running locally (seems the bundled ffi and sassc-rails don't play nicely with MRI 2.5.0 - at least on my machine - and sass-rails complains about some of the variables syntax. Which version of Ruby are you using? Though I see Travis is happy enough getting that far…)

Using sass-rails I can at least run a Rails console though, and my tests with your branch are working fine (though my tests are certainly pretty simple)… in what way is `tag_list` not working? Certainly keen to reproduce the issue in full locally.

> I actually don't see a need to call `super` here. What was the reason for this?

Prior to 2.2.0, taggable models didn't include `tag_names` in their list of model changes (i.e. `location.changes` etc) - and part of making this work in modern versions of Rails meant calling through to the underlying method implementation provided by [the `attribute` class method](https://github.com/pat/gutentag/blob/master/lib/gutentag/active_record.rb#L37). Previously, `tag_names` was just stored as a simple array in an instance variable.

**Comment by pat** [comment]
Oh, I see, the issue's more in `tag_names` than `tag_names=`. Hmm, okay, will investigate further.

**Comment by pat** [comment]
Okay, I think I've got a fix sorted. Can you try the latest in the `master` branch to confirm (and if it's all green for you, then I'll get a new gem release out):

```ruby
gem 'gutentag', '~> 2.2.0',
  git:    'https://github.com/pat/gutentag.git',
  branch: 'master'
```

**Comment by pat** [comment]
BTW, this fix should ensure `tag_names` works, and you should also be able to override `tag_names=` with your custom behaviour as well :)

**Comment by tvdeyen** [comment]
@pat it's working! Thanks for the help.

**Comment by pat** [comment]
Excellent - I just released v2.2.1 with the fix.
